flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/history/HistoryServerTest.java

```diff
@@ -117,45 +124,145 @@ public void testHistoryServerIntegration() throws Exception {
-		historyServerConfig.setLong(HistoryServerOptions.HISTORY_SERVER_ARCHIVE_REFRESH_INTERVAL, 100L);
+		CountDownLatch numExpectedArchivedJobs = new CountDownLatch(numJobs + numLegacyJobs);
 
-		historyServerConfig.setInteger(HistoryServerOptions.HISTORY_SERVER_WEB_PORT, 0);
+		Configuration historyServerConfig = createTestConfiguration(false);
 
-		// the job is archived asynchronously after env.execute() returns
-		File[] archives = jmDirectory.listFiles();
-		while (archives == null || archives.length != numJobs + 1) {
-			Thread.sleep(50);
-			archives = jmDirectory.listFiles();
+		HistoryServer hs =
+		new HistoryServer(
+			historyServerConfig,
+			(event) -> {
+				if (event.getType() == HistoryServerArchiveFetcher.ArchiveEventType.CREATED) {
+					numExpectedArchivedJobs.countDown();
+				}
+			});
+
+		try {
+			hs.start();
+			String baseUrl = "http://localhost:" + hs.getWebPort();
+			numExpectedArchivedJobs.await(10L, TimeUnit.SECONDS);
+
+			ObjectMapper mapper = new ObjectMapper();
+			Assert.assertEquals(numJobs + numLegacyJobs, getJobsOverview(baseUrl, mapper).getJobs().size());
+
+		} finally {
+			hs.stop();
+		}
+	}
+
+	@Test
+	public void testCleanExpiredJob() throws Exception {
+		runArchiveExpirationTest(true);
+	}
+
+	@Test
+	public void testRemainExpiredJob() throws Exception {
+		runArchiveExpirationTest(false);
+	}
+
+	private void runArchiveExpirationTest(boolean cleanupExpiredJobs) throws Exception {
+		int numExpiredJobs = cleanupExpiredJobs ? 1 : 0;
+		int numJobs = 3;
+		List<String> jobIDs = new ArrayList<>();
+		for (int x = 0; x < numJobs; x++) {
+			jobIDs.add(runJob());
 		}
+		waitForArchivesCreation(numJobs);
+
+		CountDownLatch numExpectedArchivedJobs = new CountDownLatch(numJobs);
+		CountDownLatch numExpectedExpiredJobs = new CountDownLatch(numExpiredJobs);
+
+		Configuration historyServerConfig = createTestConfiguration(cleanupExpiredJobs);
+
+		HistoryServer hs =
+			new HistoryServer(
+				historyServerConfig,
+				(event) -> {
+					switch (event.getType()){
+						case CREATED:
+							numExpectedArchivedJobs.countDown();
+							break;
+						case DELETED:
+							numExpectedExpiredJobs.countDown();
+							break;
+					}
+				});
 
-		HistoryServer hs = new HistoryServer(historyServerConfig, numExpectedArchivedJobs);
 		try {
 			hs.start();
+			// check that expected count of archives exists
 			String baseUrl = "http://localhost:" + hs.getWebPort();
 			numExpectedArchivedJobs.await(10L, TimeUnit.SECONDS);
 
 			ObjectMapper mapper = new ObjectMapper();
-			String response = getFromHTTP(baseUrl + JobsOverviewHeaders.URL);
-			MultipleJobsDetails overview = mapper.readValue(response, MultipleJobsDetails.class);
+			Assert.assertEquals(numJobs, getJobsOverview(baseUrl, mapper).getJobs().size());
+
+			String jobIDToDelete = jobIDs.get(0);
+			verifyArchiveFilesExistence(jobIDToDelete, true);
+
+			// delete one archive from jm
+			URI resolve = jmDirectory.toURI().resolve(jobIDToDelete);
```

Comment:
same as below, conversion to `Path` are more common

Reply:
Fixed, thanks